File: MatLab/DPR_demo.m
%% Start the parallel computing
clear,clc
gcp

%% Add path and direction of the images
addpath(genpath('DPR_function'))
data_folder = 'Test_image'; % folder  where all the files are located.
filetype = 'tif'; % type of files to be processed

%% Set image parameter
file_name = 'sarcomere';

% Determine the number of frames in the image stack
image_info = imfinfo(fullfile(data_folder, [file_name, '.', filetype]));
n = numel(image_info); % Automatically determine the number of frames

%% Load the initial TIFF image
tiff_file_path = fullfile(data_folder, [file_name, '.', filetype]);
initial_image = imread(tiff_file_path, 'Index', 1); % Load the first frame

%% Prepare image for saving
save_folder = 'DPR_image'; % folder where all the DPR-enhanced images are saved.
if ~exist(save_folder, 'dir')
    mkdir(save_folder);
end

%% Set DPR parameters

% PSF FWHM in pixels, background is the radius of the local-minimum filter in pixels, temporal analysis
% If PSF is unknown, its FWHM in pixels can be simply estimated by (0.61*lambda/NA)/pixel_size. 
% lambda is the emission wavelength, NA is the numerical aperture of the objective
PSF = 4;
options = DPRSetParameters(PSF,'gain',2,'background',10,'temporal','mean'); 

%% Load the image
img = [];
% Input image requires the DOUBLE data type
parfor i = 1:n
%     % run on Windows
%     img(:,:,i) = double(imread([data_folder,'\',file_name,'.',filetype],i)); 
    % run on Macbook
    img(:,:,i) = double(imread([data_folder,'/',file_name,'.',filetype],i)); 
end

%% Run DPR

% Input image requires the DOUBLE data type
% Output the DPR-enhanced image and the magnified raw images for comparison
[I_DPR,raw_magnified] = DPRStack(img,PSF,options);

% % For single-frame image
% [I_DPR,raw_magnified] = DPR_UpdateSingle_mex(img,PSF,options);

save_tiff_img(I_DPR,save_folder,[file_name, '_DPR2'])
cd ..
save_tiff_img(mean(raw_magnified,3),save_folder,[file_name, '_magnified'])

%% Display the initial, magnified, result images for comparison
figure
subplot(1,3,1);
imagesc(initial_image)
colormap gray
title('Initial');

subplot(1,3,2)
imagesc(mean(raw_magnified,3))
colormap gray
title("DprMagnified")

subplot(1,3,3)
imagesc(I_DPR)
colormap gray
title("DprResult")

% Set a reasonable width for the figure window
set(gcf, 'Position', [100, 100, 1200, 400]); % Adjust as needed for better visualization

File: dpr_python/dpr_demo.py
import matplotlib.pyplot as plt
import numpy as np
from dpr_function import dpr_set_parameters, process_image


def get_user_input(prompt, default=None):
    """Get user input with a default value."""
    response = input(f"{prompt} [{default}]: ")
    return response if response else default


def display_images(initial_image, magnified_image, result_image):
    """Display the initial, magnified, and result images for comparison."""
    plt.figure(figsize=(12, 4))

    plt.subplot(1, 3, 1)
    plt.imshow(initial_image[..., 0], cmap='gray')
    plt.title('Initial')

    plt.subplot(1, 3, 2)
    plt.imshow(np.mean(magnified_image, axis=2), cmap='gray')
    plt.title('DPR_Magnified')

    plt.subplot(1, 3, 3)
    plt.imshow(result_image, cmap='gray')
    plt.title('DPR_Result')

    plt.tight_layout()
    plt.show()


def main():
    # Get user inputs
    data_folder = 'test_images'
    file_name_with_type = get_user_input(
        "Enter the file name with type or press Enter to use the default value:",
        "test_image.tif"
    )
    file_name, file_type = file_name_with_type.rsplit('.', 1)

    # Ask user if they want to use default parameters
    use_default_params = get_user_input(
        "Do you want to use default parameters (psf=4, gain=2, background=10, temporal='mean')? (y/n) \n"
        "Press Enter to use the default value:",
        "y"
    ).lower()

    if use_default_params == 'y':
        psf = 4
        gain = 2
        background = 10
        temporal = 'mean'
    else:
        print("Enter the values for the following parameters:")
        psf = float(get_user_input("PSF (Point Spread Function) - Determines the blur radius", "4"))
        gain = float(get_user_input("Gain - Controls the intensity enhancement", "2"))
        background = float(get_user_input("Background - Sets the background level to subtract", "10"))
        temporal = get_user_input("Temporal - Temporal analysis method ('mean' or 'var')", "mean")

    options = dpr_set_parameters.dpr_set_parameters(psf, gain=gain, background=background, temporal=temporal)

    # Run the demo
    initial_image, dpr_image, magnified_image = process_image.process_image(data_folder, file_name, file_type, psf, options)

    if initial_image is not None:
        # Display the images
        display_images(initial_image, magnified_image, dpr_image)


if __name__ == '__main__':
    main()
